### PR TITLE
update function apps and app slots to correctly maintain content settings

### LINKED ISSUE #15855: [BUG] azurerm_windows_function_app removes Content Settings when updating resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request

### Terraform (and AzureRM Provider) Version

+ Terraform v1.1.7on linux_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.99.0

### Terraform Configuration Files

```hcl
provider "azurerm" {
  features {}
  skip_provider_registration = true
}

variable resource_group_name {}

data "azurerm_resource_group" "example" {
  name = var.resource_group_name
}

resource "random_string" "suffix" {
  length  = 12
  lower   = true
  number  = true
  upper   = false
  special = false
}

resource "azurerm_storage_account" "fx" {
  name                     = "st${random_string.suffix.result}"
  resource_group_name      = data.azurerm_resource_group.example.name
  location                 = data.azurerm_resource_group.example.location
  account_tier             = "Standard"
  account_replication_type = "LRS"
}

resource "azurerm_service_plan" "example" {
  name                = "plan-${random_string.suffix.result}"
  resource_group_name = data.azurerm_resource_group.example.name
  location            = data.azurerm_resource_group.example.location
  os_type             = "Windows"
  sku_name            = "Y1"
}

resource "azurerm_windows_function_app" "example" {
  name                = "func-${random_string.suffix.result}"
  resource_group_name = data.azurerm_resource_group.example.name
  location            = data.azurerm_resource_group.example.location

  storage_account_name = azurerm_storage_account.fx.name
  storage_account_access_key = azurerm_storage_account.fx.primary_access_key
  service_plan_id      = azurerm_service_plan.example.id

  app_settings = {
    # Uncomment the following two lines after the initial apply and run
    # terraform apply again:
    # # Disable default landing page:
    # AzureWebJobsDisableHomepage = true
  }

  site_config {}
}

```

### Description / Feedback

After the first terraform apply, you'll notice the provider has added the WEBSITE_CONTENT app settings as expected:
```JSON
❯ az functionapp config appsettings list -n func-m1cwod7cvhlb -g REDACTED
[
  {
    "name": "FUNCTIONS_EXTENSION_VERSION",
    "slotSetting": false,
    "value": "~4"
  },
  {
    "name": "AzureWebJobsStorage",
    "slotSetting": false,
    "value": "DefaultEndpointsProtocol=https;AccountName=stm1cwod7cvhlb;AccountKey=REDACTED;EndpointSuffix=core.windows.net"
  },
  {
    "name": "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING",
    "slotSetting": false,
    "value": "DefaultEndpointsProtocol=https;AccountName=stm1cwod7cvhlb;AccountKey=REDACTED;EndpointSuffix=core.windows.net"
  },
  {
    "name": "AzureWebJobsDashboard",
    "slotSetting": false,
    "value": "DefaultEndpointsProtocol=https;AccountName=stm1cwod7cvhlb;AccountKey=REDACTED;EndpointSuffix=core.windows.net"
  },
  {
    "name": "WEBSITE_CONTENTSHARE",
    "slotSetting": false,
    "value": "func-m1cwod7cvhlb-056b"
  }
]
```

Subsequent terraform apply's that include a change to the app settings, will erroneously remove the WEBSITE_CONTENT* app settings from the resource. E.g. uncomment the `AzureWebJobsDisableHomepage = true` config line and run terraform apply again:
```bash
❯ terraform apply -var 'resource_group_name=REDACTED'
random_string.suffix: Refreshing state... [id=m1cwod7cvhlb]
azurerm_service_plan.example: Refreshing state... [id=/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Web/serverfarms/plan-m1cwod7cvhlb]
azurerm_storage_account.fx: Refreshing state... [id=/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Storage/storageAccounts/stm1cwod7cvhlb]
azurerm_windows_function_app.example: Refreshing state... [id=/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Web/sites/func-m1cwod7cvhlb]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # azurerm_windows_function_app.example will be updated in-place
  ~ resource "azurerm_windows_function_app" "example" {
      ~ app_settings                      = {
          + "AzureWebJobsDisableHomepage" = "true"
        }
        id                                = "/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Web/sites/func-m1cwod7cvhlb"
        name                              = "func-m1cwod7cvhlb"
        tags                              = {}
        # (23 unchanged attributes hidden)


        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

azurerm_windows_function_app.example: Modifying... [id=/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Web/sites/func-m1cwod7cvhlb]
azurerm_windows_function_app.example: Still modifying... [id=/subscriptions/REDACTED-.../Microsoft.Web/sites/func-m1cwod7cvhlb, 10s elapsed]
azurerm_windows_function_app.example: Modifications complete after 1m2s [id=/subscriptions/REDACTED/resourceGroups/REDACTED/providers/Microsoft.Web/sites/func-m1cwod7cvhlb]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```
Inspection of the app settings reveals that the WEBSITE_CONTENT settings have been removed (even though the plan didn't indicate they would in fact be removed):
```json
❯ az functionapp config appsettings list -n func-m1cwod7cvhlb -g REDACTED
[
  {
    "name": "FUNCTIONS_EXTENSION_VERSION",
    "slotSetting": false,
    "value": "~4"
  },
  {
    "name": "AzureWebJobsStorage",
    "slotSetting": false,
    "value": "DefaultEndpointsProtocol=https;AccountName=stm1cwod7cvhlb;AccountKey=REDACTED;EndpointSuffix=core.windows.net"
  },
  {
    "name": "AzureWebJobsDisableHomepage",
    "slotSetting": false,
    "value": "true"
  },
  {
    "name": "AzureWebJobsDashboard",
    "slotSetting": false,
    "value": "DefaultEndpointsProtocol=https;AccountName=stm1cwod7cvhlb;AccountKey=REDACTED;EndpointSuffix=core.windows.net"
  }
]
```

For some function plans the [WEBSITE_CONTENT](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#website_contentazurefileconnectionstring)* settings are required, otherwise its function(s) fail to execute. This unwanted removal of the content settings is especially problematic as they are set correctly when the resource is created, however any subsequent update of the resource silently removes the settings; making the function app fail in an unapparent way.

Final note: other resources besides `azurerm_windows_function_app` might suffer from the same bug, the [azure docs](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#website_contentazurefileconnectionstring) state that these website content settings are mandatory for `Windows or Linux Premium plan or to a Windows Consumption plan`.

### References

This issue was reported before in #9766. The bug in this issue is very similar, but here it is filled against the beta resources.

### PR DESCRIPTION
**Pull request by jackofallops**
fixes #15855

* `azurerm_linux_function_app` - fixed update handling of `app_settings` for `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING`
* `azurerm_linux_function_app_slot` - fixed update handling of `app_settings` for `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING`
* `azurerm_windows_function_app` - fixed update handling of `app_settings` for `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING`
* `azurerm_windows_function_app_slot` - fixed update handling of `app_settings` for `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING`

Tests passing:
image

### CODE REVIEW

File: internal/services/appservice/linux_function_app_slot_resource.go [hunk 1]
```diff
@@ -655,6 +655,13 @@ func (r LinuxFunctionAppSlotResource) Update() sdk.ResourceFunc {
 				return fmt.Errorf("reading Linux %s: %v", id, err)
 			}
 
+			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id)
+			if err != nil {
+				return err
+			}
+
+			sendContentSettings := !helpers.PlanIsElastic(*planSKU)
```
Comment (katbyte) [inline internal/services/appservice/linux_function_app_slot_resource.go:663]: should helpers.PlanIsElastic take a nil pointer so avoid all the derefs (and then always check for nil) ?

**Review by katbyte — APPROVED** [review]
aside from one minor comment LGTM 🏗️

**Comment by fvdnabee** [comment]
Verified that this PR (commit 85dc4c054) keeps the content settings after a resource update (for azurerm_windows_function_app). Thanks jack!

**Review by katbyte — APPROVED** [review]
LGTM! 👍

**Pull request merged**